posts/2018-06-15-typedapi.md

```diff
@@ -0,0 +1,329 @@
+
+#### Type-level FoldLeft
+What we want is to go from `Api <: HList` to `(El <: HList, KIn <: HList, VIn <: HList, M, Out)` with:
+  * `El` al the elements in our api: `"users".type :: SegmentInput :: QueryInput :: GetCall :: HNil`
+  * `KIn` the input key types: `nameW.T :: minAgeW.T :: HNil`
+  * `VIn` the input value types: `String :: Int :: HNil`
+  * the method type: `GetCall`
+  * and `Out`: `List[User]`
+
+Here, we introduced new types `SegmentInput` and `QueryInput` which act as placeholders and indicate that our api has the following inputs. This representation will come in handy when we construct our function.
+
+Now, how to fold on the type-level? The first step, we have to define a function which describes how to aggregate two types:
+
+```scala
+trait FoldLeftFunction[In, Agg] { type Out }
+```
+
+That's it. We say what goes in and what comes out. You need some examples to get a better idea? Here you go:
+
+```scala
+implicit def pathTransformer[P, El <: HList, KIn <: HList, VIn <: HList, M, Out] = 
+  FoldLeftFunction[Path[P], (El, KIn, VIn, M, Out)] { type Out = (P :: El, KIn, VIn, Out) }
+
+implicit def segmentTransformer[K <: Symbol, V, El <: HList, KIn <: HList, VIn <: HList, M, Out] = 
+  FoldLeftFunction[Segment[K, V], (El, KIn, VIn, M, Out)] { type Out = (SegmentInput :: El, K :: KIn, V :: VIn, Out) }
+
+// and so on
+```
+
+Now that we can aggregate types we need a vehicle to traverse our `HList` type and transform it on the fly by using our `FoldLeftFunction` instances. I think yet another type class can help us here.
+
+```scala
+trait TypeLevelFoldLeft[H <: HList, Agg] { type Out }
+
+implicit def returnCase[Agg] = new TypeLevelFoldLeft[HNil, Agg] {
+  type Out = Agg
+}
+
+implicit def foldCase[H, T <: HList, Agg, FfOut, FOut](implicit f: FoldLeftFunction.Aux[H, Agg, FfOut], 
+                                                                next: Lazy[TypeLevelFoldLeft.Aux[T, FfOut, FOut]]) = 
+  new TypeLevelFoldLeft[H :: T, Agg] { type Out = FOut }
+```
+
+The above definition describes a recursive function which will apply the `FoldLeftFunction` on `H` and the current aggregated type `Agg` and continues with the resulting `FfOut` and the remaining list. And before you bang your head against the wall for hours until the clock strikes 3 am, like I did, a small hint, make `next` lazy. Otherwise, Scala is not able to find `next`. My guess is that Scala is not able to infer `next`, because it depends on `FfOut` which is also unknown. So we have to defer `next`'s inference to give the compiler some time to work.
+
+And another hint, you can start with `Nothing` as the initial type for your aggregate.
+
+#### Collect all the request data
+We folded our api type into the new representation. Now we use that to build a function which collects all the data necessary to make a request.
+
+```scala
+type Uri     = List[String]
+type Queries = Map[String, List[String]]
+
```

Comment:
again, a short description in English of what these types represent might be helpful, especially stressing the fact that these live in the value level (that isn't a correct way to put it, but hopefully you see what I mean).